qa/L0_implicit_state/implicit_state.py

```diff
@@ -0,0 +1,135 @@
+#  * Neither the name of NVIDIA CORPORATION nor the names of its
+#    contributors may be used to endorse or promote products derived
+#    from this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+# PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import sys
+
+sys.path.append('../common')
+
+import argparse
+import numpy as np
+import os
+from builtins import range
+import tritonclient.http as tritonhttpclient
+from tritonclient.utils import InferenceServerException
+import unittest
+import test_util as tu
+BACKENDS = os.environ.get('BACKENDS', "onnx plan")
+
+
+class ImplicitStateTest(tu.TestResultCollector):
+
+    def test_no_implicit_state(self):
+        triton_client = tritonhttpclient.InferenceServerClient("localhost:8000")
+        inputs = []
+        inputs.append(tritonhttpclient.InferInput('INPUT', [1], 'INT32'))
+        inputs.append(tritonhttpclient.InferInput('TEST_CASE', [1], 'INT32'))
+        inputs[0].set_data_from_numpy(np.random.randint(5, size=[1], dtype=np.int32))
+        inputs[1].set_data_from_numpy(np.asarray([0], dtype=np.int32))
+
+        with self.assertRaises(InferenceServerException) as e:
+            triton_client.infer(model_name="no_implicit_state", inputs=inputs, sequence_id=1, sequence_start=True)
+
+        self.assertEqual(str(e.exception), "unable to add state 'undefined_state'. State configuration is missing for model 'no_implicit_state'.")
+
+    def test_wrong_implicit_state_name(self):
+        triton_client = tritonhttpclient.InferenceServerClient("localhost:8000")
+        inputs = []
+        inputs.append(tritonhttpclient.InferInput('INPUT', [1], 'INT32'))
+        inputs.append(tritonhttpclient.InferInput('TEST_CASE', [1], 'INT32'))
+        inputs[0].set_data_from_numpy(np.random.randint(5, size=[1], dtype=np.int32))
+        inputs[1].set_data_from_numpy(np.asarray([0], dtype=np.int32))
+
+        with self.assertRaises(InferenceServerException) as e:
+            triton_client.infer(model_name="wrong_internal_state", inputs=inputs, sequence_id=2, sequence_start=True)
+
+        self.assertEqual(str(e.exception), "state 'undefined_state' is not a valid state name.")
+
+    def test_no_update(self):
+	    # Test implicit state without updating any state
```

Comment:
Can you elaborate what this test is doing? Why the implicit state won't be updated when it receives a sequence of requests? Is this just how the model work? i.e. it always output the same data for the "state".

Reply:
I added this test to make sure that `TRITONBACKEND_StateUpdate` must be called to update the state. In the phase 2 of implicit state, there is a feature request to add the capability to not update some state based on some inputs (cloning of implicit state ticket). I wanted to make sure that this is behaving as expected.

Reply:
Are you saying that `TRITONBACKEND_StateUpdate` must NOT be called?

Reply:
No, I wanted to make sure that not calling it will lead to the state not being updated. It should be called if you want to update the internal state.